test/integration/controllermanager/shootstate/shootstate_test.go

```diff
@@ -0,0 +1,240 @@
+				addFinalizer(shootState)
+
+				By("Update the Shoot's last operation")
+				Expect(testClient.Status().Patch(ctx, shoot, patch)).To(Succeed())
+
+				By("Verify the finalizer is present")
+				Consistently(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(shootState), shootState)).To(Succeed())
+					return shootState.Finalizers
+				}).Should(Equal([]string{shootstate.FinalizerName}))
+			})
+		})
+
+		When("Shoot last operation has Restore type", func() {
+			BeforeEach(func() {
+				shoot.Status.LastOperation = &gardencorev1beta1.LastOperation{
+					Type:     gardencorev1beta1.LastOperationTypeRestore,
+					State:    gardencorev1beta1.LastOperationStateProcessing,
+					Progress: 0,
+				}
+			})
+
+			It("should add finalizer if not present and operation has not succeeded", func() {
+				By("Update the Shoot's last operation")
+				Expect(testClient.Status().Patch(ctx, shoot, patch)).To(Succeed())
+
+				By("Verify the finalizer is present")
+				Eventually(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(shootState), shootState)).To(Succeed())
+					return shootState.Finalizers
+				}).Should(ConsistOf(shootstate.FinalizerName))
+			})
+
+			It("should not duplicate finalizer if already present", func() {
+				addFinalizer(shootState)
+
+				By("Update the Shoot's last operation")
+				Expect(testClient.Status().Patch(ctx, shoot, patch)).To(Succeed())
+
+				By("Verify the finalizer is present")
+				Consistently(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(shootState), shootState)).To(Succeed())
+					return shootState.Finalizers
+				}).Should(Equal([]string{shootstate.FinalizerName}))
+			})
+
+			It("should remove finalizer if present and operation has succeeded", func() {
+				addFinalizer(shootState)
+
+				By("Update the Shoot's last operation")
+				shoot.Status.LastOperation.State = gardencorev1beta1.LastOperationStateSucceeded
+				Expect(testClient.Status().Patch(ctx, shoot, patch)).To(Succeed())
+
+				By("Verify the finalizer is removed")
+				Eventually(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(shootState), shootState)).To(Succeed())
+					return shootState.Finalizers
+				}).ShouldNot(ConsistOf(shootstate.FinalizerName))
+			})
+
+			It("should not remove finalizer if present and operation has not succeeded ", func() {
+				addFinalizer(shootState)
+
+				By("Update the Shoot's last operation")
+				Expect(testClient.Status().Patch(ctx, shoot, patch)).To(Succeed())
+
+				By("Verify the finalizer is present")
+				Consistently(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(shootState), shootState)).To(Succeed())
+					return shootState.Finalizers
+				}).Should(Equal([]string{shootstate.FinalizerName}))
+			})
+		})
+
+		When("Shoot last operation has Reconcile type", func() {
```

Comment:
```suggestion
		When("Shoot last operation is Reconcile", func() {
```

Reply:
Thanks for the suggestion, reused it in my latest updates.